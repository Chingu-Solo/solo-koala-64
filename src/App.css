.Card {
  display: flex;
  flex-direction: column;
  justify-content: center;
  font-size: 14px;
  color: #000;
}

.Cards {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  font-size: 14px;
  color: #000;
}

.Tools {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  font-size: 14px;
  color: #000;
}


.App {
  text-align: center;
  color: #000;
  background-color: #fff;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

header {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

footer {
  width: 100%;
  text-align: center;
  font-size: 10px;
  position: absolute;
  bottom: 10px;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
