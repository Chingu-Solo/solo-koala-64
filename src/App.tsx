import React, { forwardRef, Fragment } from 'react';
//import ReactDOM from "react-dom";
import './App.css';
import CSS from 'csstype';

import { 
  FixedSizeList as List,
  FixedSizeGrid as Grid
} from 'react-window';
import { FaList, FaRegWindowClose } from 'react-icons/fa';
import { FiGrid } from 'react-icons/fi';
import { TiArrowUp } from 'react-icons/ti';
import { AiOutlineReload, AiOutlineSelect } from 'react-icons/ai';
import AutoSizer from 'react-virtualized-auto-sizer';
//import { fonts } from './fixtures/fonts';
import { TextInput } from './components/Tools';
import GoogleFontsAPI, { GoogleFont } from './api/GoogleFonts';
import { EventIdHandler } from './common/types';

type Input = string;

interface CardBase {
  input: Input,
  clickRemoveHandler: EventIdHandler
  clickToTopHandler: EventIdHandler
}

interface Data extends CardBase {
  fonts: GoogleFont[],
}

interface ReactWindowComp {
  // react-window requires to include all information in one {data} prop
  data: Data,
}

interface CardProps extends CardBase { 
  font: GoogleFont,
  listId: number
}

function Card({ 
  font, 
  input, 
  clickRemoveHandler, 
  clickToTopHandler, 
  listId 
}: CardProps) {
  const defaultText: string = (
    'Then came the night of the first falling star'
  );
  return(
    <div className="Card">
      <div className="Tools">
        <button 
          onClick={() => {}}
          title="Select this font"
        >
          <AiOutlineSelect /> 
        </button>
        <button 
          onClick={() => clickToTopHandler(listId)}
          title="Move font to top"
        >
          <TiArrowUp /> 
        </button>
        <button 
          onClick={() => clickRemoveHandler(listId)}
          title="Remove Font (! New Http Request needed to redo)"
        >
          <FaRegWindowClose /> 
        </button>
      </div>
      <p>{font.family}</p>
      <div>
        <link rel="stylesheet" type="text/css" href={font.styleSheetURL} />
        <p style={{fontFamily: `'${font.family}', serif`}}>
          {input || defaultText}
        </p>
      </div>
    </div>    
  );
}


interface StyledCardsContainer extends ReactWindowComp {
  style: CSS.Properties | any,
}

interface RowProps extends StyledCardsContainer {
  //TODO typing more specific ?
  index?: number,
}

function Row({ index=0, style, data }: RowProps) {
  return (
    <div style={style}>
      <Card 
        font={data.fonts[index]} 
        input={data.input} 
        clickRemoveHandler={data.clickRemoveHandler}
        clickToTopHandler={data.clickToTopHandler}
        listId={index}
      />
    </div>
  );
};


const GUTTER_SIZE = 5;
const COLUMN_WIDTH = 250;
const ROW_HEIGHT = 100;
const COLUMN_COUNT = 3;

interface CellProps extends StyledCardsContainer {
  columnIndex: number, 
  rowIndex: number, 
}


function Cell({ columnIndex, rowIndex, style, data }: CellProps) {
  const listIndex = columnIndex*COLUMN_COUNT + rowIndex;
  const i: number = Math.min(listIndex, data.fonts.length-1);
  const displayI: boolean = listIndex < data.fonts.length;
  return (
    <div
      className={"GridItem"}
      style={{
        ...style,
        left: style.left + GUTTER_SIZE,
        top: style.top + GUTTER_SIZE,
        width: style.width - GUTTER_SIZE,
        height: style.height - GUTTER_SIZE
      }}
    >
      {displayI &&
        <Card 
          font={data.fonts[i]} 
          input={data.input} 
          clickRemoveHandler={data.clickRemoveHandler}
          clickToTopHandler={data.clickToTopHandler}
          listId={i}
        />
      }
    </div>
  );
}

const innerElementType = forwardRef(({ style, ...rest }: CellProps, ref: any) => (
  <div
    ref={ref}
    style={{
      ...style,
      paddingLeft: GUTTER_SIZE,
      paddingTop: GUTTER_SIZE
    }}
    {...rest}
  />
));


type CardsDisplay = 'grid' | 'list';

interface CardsProps extends ReactWindowComp { 
  cardsDisplay: CardsDisplay,
}

class Cards extends React.Component<CardsProps> {
  //TODO Font Name, the sample text, and an add button
  listRef: any = React.createRef();
  gridRef: any = React.createRef();
  render () {
    return(
      <Fragment>
        <div>
        <button 
          onClick={() => {
            if (this.props.cardsDisplay === 'list') {
              this.gridRef.current.scrollToItem({
                align: "start",
                columnIndex: 0,
                rowIndex: 0,
              });
            } else {
              this.listRef.current.scrollToItem(0, 'start');
            }
          }}
          title="Scroll to top"
        >
          <TiArrowUp /> 
        </button>
      </div>
        <AutoSizer>
           {this.props.cardsDisplay === 'list'
             ? ({ height, width }) => (
                <Grid
                  className="Grid"
                  columnCount={COLUMN_COUNT}
                  columnWidth={COLUMN_WIDTH + GUTTER_SIZE}
                  height={height}
                  innerElementType={innerElementType}
                  rowCount={Math.ceil(this.props.data.fonts.length/COLUMN_COUNT)}
                  rowHeight={ROW_HEIGHT + GUTTER_SIZE}
                  width={width}
                  itemData={this.props.data}
                  ref={this.gridRef}
                >
                  {Cell}
                </Grid>
               )
             : ({ height, width }) => (
               <List
                  className="List"
                  height={height}
                  itemCount={this.props.data.fonts.length}
                  itemSize={150} // itemSize
                  width={width}
                  itemData={this.props.data}
                  ref={this.listRef}
                >
                  {Row}
                </List>
              )
          } 
        </AutoSizer>
      </Fragment>
    );
  }
}

type Fonts = GoogleFont[] | null;
interface AppState {
  fontsAPI: GoogleFontsAPI,
  fonts: Fonts,
  bakFonts: Fonts,
  cardsDisplay: CardsDisplay,
  inputText: Input,
  searchText: string
}

export default class App extends React.Component {
  readonly state: AppState = {
    fontsAPI: new GoogleFontsAPI(),
    fonts: null,  //or maybe set some default
    bakFonts: null, // backup to reset without request after messing with state
    cardsDisplay: 'list',
    inputText: '',
    searchText: ''
  }

  async componentDidMount() {
    const fonts: GoogleFont[] = await this.state.fontsAPI._getGoogleFonts();
    this.setState({ fonts })
    fonts && this.setState({ bakFonts: [...fonts] })
  }

  filteredFonts(): Fonts {
    const searchText = this.state.searchText;
    const fonts = this.state.fonts;
    if (!searchText) {
      return fonts
    } else {
      return (
        fonts && fonts.filter(
          font => font.family.toLowerCase().includes(searchText.toLowerCase())
        )
      );
    }
  }

  resetState = () => this.setState({ 
    cardsDisplay: 'list',
    searchText: '',
    inputText: '',
    fonts: this.state.bakFonts, //because we messed with state
  });

  setListOrGrid = () => this.setState({
    cardsDisplay: this.state.cardsDisplay === 'list' ? 'grid': 'list'
  });

  render() {
    const fonts = this.filteredFonts();
    // all fonts stay in the state, and no new API request needed
    return (
      <div className="App">
        <header>
          <p>Google Fonts</p>
        </header>
        <div className="Tools">
          <TextInput
            value={this.state.searchText}
            placeHolder="search fonts" 
            changeHandler={e => this.setState({ searchText: e.target.value })}
          />
          <TextInput 
            value={this.state.inputText}
            placeHolder="type-something" 
            changeHandler={e => this.setState({ inputText: e.target.value })}
          />
          <button 
            onClick={this.setListOrGrid}
            title={`View as ${this.state.cardsDisplay}`}
          >
            {this.state.cardsDisplay === 'list'
              ? <FaList />
              : <FiGrid />
            }
          </button>
          <button 
            onClick={this.resetState}
            title="Reset"
          >
            <AiOutlineReload />
          </button>
        </div>
        <div className="Cards">
          {fonts && 
            <Cards 
              data={{
                fonts, 
                input: this.state.inputText,
                clickRemoveHandler: listId => {
                  //if (sureFonts.indexOf(listId) !== -1) { //TODO type problem
                  if (fonts.length > listId) {
                    fonts.splice(listId, 1);
                    this.setState({ fonts: fonts });
                  }
                },
                clickToTopHandler: listId => {
                  if (fonts.length > listId) {
                    const newTop = fonts.splice(listId, 1)[0];
                    this.setState({ fonts: [newTop, ...fonts]});
                  }
                }
              }} 
              cardsDisplay={this.state.cardsDisplay}
            />
          }
        </div>
        <footer>
          <p>coded by faebebin | 2020 | Chingu Pre-Work Project</p>
        </footer>
      </div>
    );
  }
}
