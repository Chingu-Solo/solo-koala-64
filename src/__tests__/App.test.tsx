import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import App from '../App';

describe('App', () => {

  it('renders footer text', () => {
    // maybe add some more ;) ...
    const { getByText } = render(<App />);
    const linkElement = getByText(/faebebin/i);
    expect(linkElement).toBeInTheDocument();
  });

  it(`renders screen as expected`, () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('fetches data from server when server returns a successful response', done => {
    const API_KEY = process.env.REACT_APP_GOOGLE_FONTS_API_KEY;
    const request: string = `https://www.googleapis.com/webfonts/v1/webfonts?key=${API_KEY}&sort=popularity`
    const mockSuccessResponse = {} //new Response();
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockSuccessResponse);
    
    const wrapper: any = render(<App />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(request);

      done();
      //    process.nextTick(() => {
      //      expect(wrapper.sta).toEqual({
      //        // ... assert the set state
      //      });
      //
      //      global.fetch.mockClear();
      //    });
  });
});

