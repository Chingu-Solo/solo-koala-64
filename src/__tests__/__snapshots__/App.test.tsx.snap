// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders screen as expected 1`] = `
<div
  className="App LightBlack"
>
  <nav
    className="LightBlack navbar navbar-expand-md navbar-light sticky-top"
  >
    <a
      className="navbar-brand"
      href="#home"
    >
      Google Fonts
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="mr-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key="#"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          Catalog
        </a>
        <a
          className="nav-link"
          data-rb-event-key="#"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          Featured
        </a>
        <a
          className="nav-link"
          data-rb-event-key="#"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          Articles
        </a>
        <a
          className="nav-link"
          data-rb-event-key="#"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          About
        </a>
      </div>
    </div>
  </nav>
  <div
    className="Tools MainToolBar"
  >
    <input
      className="TextInput LightBlack"
      onChange={[Function]}
      placeholder="🔍 search fonts"
      type="text"
      value=""
    />
    <input
      className="TextInput LightBlack"
      onChange={[Function]}
      placeholder="✎ type-something"
      type="text"
      value=""
    />
    <select
      className="LightBlack"
      onChange={[Function]}
      title="chose font size"
    >
      <option
        value="8px"
      >
        8px
      </option>
      <option
        value="12px"
      >
        12px
      </option>
      <option
        value="14px"
      >
        14px
      </option>
      <option
        value="20px"
      >
        20px
      </option>
      <option
        value="24px"
      >
        24px
      </option>
      <option
        value="32px"
      >
        32px
      </option>
      <option
        value="40px"
      >
        40px
      </option>
    </select>
    <select
      className="LightBlack buttonIcon"
      onChange={[Function]}
      title="chose Color"
    >
      <option
        className="LightBlack NoSelectMarker buttonIcon"
        value="LightBlack"
      >
        ◉
      </option>
      <option
        className="LightBlue NoSelectMarker buttonIcon"
        value="LightBlue"
      >
        ◉
      </option>
      <option
        className="LightGreen NoSelectMarker buttonIcon"
        value="LightGreen"
      >
        ◉
      </option>
      <option
        className="DarkWhite NoSelectMarker buttonIcon"
        value="DarkWhite"
      >
        ◉
      </option>
      <option
        className="DarkYellow NoSelectMarker buttonIcon"
        value="DarkYellow"
      >
        ◉
      </option>
      <option
        className="DarkGreen NoSelectMarker buttonIcon"
        value="DarkGreen"
      >
        ◉
      </option>
    </select>
    <button
      className="LightBlack"
      onClick={[Function]}
      title="shuffle font cards"
    >
      <svg
        baseProfile="tiny"
        className="buttonIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 9h3.5c.736 0 1.393.391 1.851 1.001.325-.604.729-1.163 1.191-1.662-.803-.823-1.866-1.339-3.042-1.339h-3.5c-.553 0-1 .448-1 1s.447 1 1 1zM11.685 12.111c.551-1.657 2.256-3.111 3.649-3.111h1.838l-1.293 1.293c-.391.391-.391 1.023 0 1.414.195.195.451.293.707.293s.512-.098.707-.293l3.707-3.707-3.707-3.707c-.391-.391-1.023-.391-1.414 0s-.391 1.023 0 1.414l1.293 1.293h-1.838c-2.274 0-4.711 1.967-5.547 4.479l-.472 1.411c-.641 1.926-2.072 3.11-2.815 3.11h-2.5c-.553 0-1 .448-1 1s.447 1 1 1h2.5c1.837 0 3.863-1.925 4.713-4.479l.472-1.41zM15.879 13.293c-.391.391-.391 1.023 0 1.414l1.293 1.293h-2.338c-1.268 0-2.33-.891-2.691-2.108-.256.75-.627 1.499-1.09 2.185.886 1.162 2.243 1.923 3.781 1.923h2.338l-1.293 1.293c-.391.391-.391 1.023 0 1.414.195.195.451.293.707.293s.512-.098.707-.293l3.707-3.707-3.707-3.707c-.391-.391-1.023-.391-1.414 0z"
        />
      </svg>
    </button>
    <button
      className="LightBlack"
      onClick={[Function]}
      title="Reset"
    >
      <svg
        className="buttonIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 0 0-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 0 1 655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 0 1 279 755.2a342.16 342.16 0 0 1-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 0 1 109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 0 0 3 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
        />
      </svg>
    </button>
  </div>
  <div>
    <div
      className="InfoBar"
    >
      Viewing 
      0
       out of 
      0
       fonts.
    </div>
  </div>
  <div
    className="CardsContainer"
  >
    <button
      className="ScrollToTop LightBlack"
      onClick={[Function]}
      title="Scroll to top"
    >
      <svg
        baseProfile="tiny"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 5.586l-4.707 4.707c-.391.391-.391 1.023 0 1.414s1.023.391 1.414 0l2.293-2.293v7.586c0 .552.447 1 1 1s1-.448 1-1v-7.586l2.293 2.293c.195.195.451.293.707.293s.512-.098.707-.293c.391-.391.391-1.023 0-1.414l-4.707-4.707z"
        />
      </svg>
    </button>
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>
  <footer>
    <p>
      coded by faebebin | 2020 | Chingu Pre-Work Project
    </p>
  </footer>
</div>
`;
